File: week2_decomposition2/3_strongly_connected/strongly_connected.py
#Uses python3

import sys

sys.setrecursionlimit(200000)

def reverse_graph(adj):
    adj_r = [[] for _ in range(len(adj))]
    for i in range(len(adj)):
        for j in adj[i]:
            adj_r[j].append(i)
    return adj_r

def number_of_strongly_connected_components(adj):
    result = 0
    #write your code here
    order = []
    used = [False]*len(adj)
    adj_r = reverse_graph(adj)

    def dfs(v,G,sort=False):
        used[v] = True
        for w in G[v]:
            if used[w]==False:
                dfs(w,G,sort)
        if sort:
            order.append(v)
    ##### toposort
    for i in range(len(adj_r)):
        if used[i]==False:
            dfs(i,adj_r,sort=True)
    ##### count ssc
    used = [False]*len(adj)
    for i in reversed(order):
        if used[i]==False:
            dfs(i,adj,sort=False)
            result += 1
    return result

if __name__ == '__main__':
    input = sys.stdin.read()
    data = list(map(int, input.split()))
    n, m = data[0:2]
    data = data[2:]
    edges = list(zip(data[0:(2 * m):2], data[1:(2 * m):2]))
    adj = [[] for _ in range(n)]
    for (a, b) in edges:
        adj[a - 1].append(b - 1)
    print(number_of_strongly_connected_components(adj))

File: week2_decomposition2/3_strongly_connected/strongly_connected.cpp
#include <algorithm>
#include <iostream>
#include <vector>
#include <functional>

using std::vector;
using std::pair;


int number_of_strongly_connected_components(vector<vector<int> > adj) {
  int result = 0;
  //write your code here
  int n = adj.size();
  vector<vector<int> > adj_r(n,vector<int>());
  vector<bool> used(n,false);
  vector<int> order(n);
  auto reverse_graph = [&](){
    for(int i=0;i<n;++i)
      for(int j: adj[i]) adj_r[j].push_back(i);
  };
  reverse_graph();

  std::function<void(int,bool,vector<vector<int> >&)> dfs=[&](int v,bool sort,vector<vector<int> >& G){
    used[v] = true;
    for(int w: G[v])
      if (used[w]==false) dfs(w,sort,G);
    if (sort) order.push_back(v);
  };
// toposort in reversed graph
  for(int i=0;i<n;++i)
    if(used[i]==false) dfs(i,true,adj_r);
  std::reverse(order.begin(),order.end());
// count ssc
  used.assign(n,false);
  for(int i:order)
    if(used[i]==false){
      dfs(i,false,adj);
      ++result;
    }

  return result;
}

int main() {
  size_t n, m;
  std::cin >> n >> m;
  vector<vector<int> > adj(n, vector<int>());
  for (size_t i = 0; i < m; i++) {
    int x, y;
    std::cin >> x >> y;
    adj[x - 1].push_back(y - 1);
  }
  std::cout << number_of_strongly_connected_components(adj);
}
